### ISSUE TITLE
Implement Kernel#Integer in Rust

### ISSUE THREAD

**Issue by lopopolo** [artichoke/artichoke#282]
`Kernel#Integer` is a complex converter from arbitrary Ruby objects to `Integer`s, and includes String radix parsing.

@tkbky worked on a pure Ruby implementation https://github.com/artichoke/artichoke/pull/254/commits/3b218c645333cea68adb37099d2db844b2cf82e8 that is pretty complicated and seems hard to exhaustively test.

Rust std has a standard library function on numeric primitives called [`from_str_radix`](https://doc.rust-lang.org/std/primitive.i64.html#method.from_str_radix) which handles all of the parsing, overflow detection, and error handling for us.

I've mocked up a parser in Rust [[playground]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=65200add357dd7d7d87b15e44db8dd62):

```rust
fn parse(s: &str, radix: Option<u32>) -> Result<i64, &'static str> {
    let mut buf = String::new();
    let mut parsed_radix = None;
    let s = if s.starts_with('-') {
        buf.push('-');
        &s[1..]
    } else {
        s
    };
    match &s[0..2] {
        "0b" | "0B" => {
            buf.push_str(&s[2..]);
            parsed_radix = Some(2);
        }
        "0o" | "0O" => {
            buf.push_str(&s[2..]);
            parsed_radix = Some(8);
        }
        "0x" | "0X" => {
            buf.push_str(&s[2..]);
            parsed_radix = Some(16);
        }
        prefix if &prefix[0..1] == "0" => {
            buf.push_str(&s[1..]);
            parsed_radix = Some(8);
        }
        _ => {}
    };
    match (radix, parsed_radix) {
        (Some(_), Some(_)) => Err("invalid value for Integer(): \"xxx\""),
        (Some(radix), None) => i64::from_str_radix(buf.as_str(), radix).map_err(|_| "invalid"),
        (None, Some(radix)) => i64::from_str_radix(buf.as_str(), radix).map_err(|_| "invalid"),
        (None, None) => i64::from_str_radix(buf.as_str(), 10).map_err(|_| "invalid"),
    }
}

fn main() {
    println!("{:?}", parse("0b111", None));
    println!("{:?}", parse("0b111", Some(5)));
}
```

**Comment by lopopolo** [comment]
@tkbky would you like to take this?

**Comment by tkbky** [comment]
@lopopolo ❤️ the PR description. Yes, I will take this.

**Comment by lopopolo** [comment]
The match can be simplified with the new `Option::xor` API available in 1.37.0.

**Comment by lopopolo** [comment]
Followup to GH-254.
